### DISCUSSION TITLE
폰트 최적화

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#59] [프론트 기술 질문 공간]
프로젝트에서 외부에서 다운받은 폰트를 사용하고자 했다. 이후 프로젝트를 잘 만들고 나니 처음 화면을 로드할 때 폰트를 받아오는 시간이 걸려서 텍스트가 보이지 않아 UX가 좋지 못하다는 피드백을 받았다.  이를 개선하기 위해서 사용할 수 있는 방식은 어떤 것이 있을까?

**Comment by michoball** [comment]
우선 생각해볼 것은 웹 폰트의 용량일 것 같습니다.   폰트를 woff2 안되면 woff 로 압축을 해서 사이즈를 줄여 다운을 빠르게하는 방법이 있겠습니다. 

local 값 적용 : @font-face 안에 src 에 local 값을 적용해서 local 에 이미 필요한 폰트가 있다면 바로 적용하는 형식입니다. 

font-display  값 적용: @font-face안에 font-display에  block, sawp, fallback등 다양한 옵션값을 사용해서 foit 나 fout 방식으로 폰트를 로드하여 사용자 경험을 좋게 만든다.

FOIT : Flash Of Unstyled Text -> 폰트 로드전 대체 폰트로 텍스트가 보임
FOUT : Flash Of Invisible Text -> 폰트 로드전 텍스트가 보이지 않음

폰트 리소스요청 앞당기기:  html head에 폰트 리소스를 요청하는 부분에  rel="preload"를 추가하는 방법입니다. 이렇게하면 폰트를 빨리 받아올 수 있지만 페이지에서 사용하지 않아도 받아온다는 단점이 있습니다.

**Comment by WooGie911** [comment]
여러 방법이 있겠지만 폰트를 받아오는 시간이 오래 걸린다는 문제이기에 용량이 작은 폰트를 사용하거나 폰트에서 불필요한 글자를 제거한  서브셋 폰트를 사용하는 방법이 있을 것 같습니다.
그치만 사용해야하는 폰트가 정해져 있다면 preload를 사용해 폰트 리소스를 초기에 요청하는 방법과, 폰트 로딩 API를 사용하여 폰트 로딩이 완료되어 사용 가능할 때 텍스트를 나타내는 방법이 있을 것 같습니다.

**Comment by javascriptbiggosoo** [comment]
사이트에서 사용하는 font-weight, font-style만 골라서 임포트 하는 것도 하나의 경우가 될 수 있을 것 같네요.
italic을 사용할 일이 없다면 가져올 필요가 없겠죠

**Comment by BaekYuri** [comment]
1. 폰트 압축률이 좋은 woff2 파일을 사용하면 로딩 시간을 줄일 수 있습니다. 만약 지원하지 않는 브라우저라면 woff를 사용할 수 있습니다.
2. 폰트 리소스는 다른 리소스들에 비해 늦게 요청되기 때문에 브라우저를 렌더링 하는 동안 폰트의 리소스 요청도 늦어집니다. 이러한 상황을 변경시켜, 폰트의 리소스 요청을 먼저 하는 방법이 있습니다.<head>에서 폰트를 요청할 때, rel=”preload”를 추가하는 방법으로 해결할 수 있습니다. 이러한 방법은 효과적일 수 있지만, 많은 양의 폰트를 로드할 때에는 오히려 오래 걸릴 수 있습니다.
3. 서브셋 폰트를 사용하는 방법이 있습니다. 한글은 조합하면 만개가 넘는 여러 글자가 만들어집니다. 그 중에서도, 사용하지 않는 글자들을 제외한 서브셋 폰트를 만들어서 사용하면 폰트의 크기를 줄일 수 있습니다.
